bluemira/magnetostatics/greens.py

```diff
@@ -288,12 +278,44 @@ def greens_dpsi_dz(
 
     The implementation used here refactors the above to avoid some zero divisions.
     """
-    a = ((x + xc) ** 2 + (z - zc) ** 2) ** 0.5
+    a, k2 = calc_a_k2(xc, zc, x, z)
+    i1, i2 = calc_i1_i2(a, k2)
+    return MU_0_2PI * ((z - zc) * (i1 - i2 * ((z - zc) ** 2 + x**2 + xc**2)))
+
+
+@nb.jit(nopython=True)
+def calc_a_k2(
+    xc: float | np.ndarray,
+    zc: float | np.ndarray,
+    x: float | np.ndarray,
+    z: float | np.ndarray,
+):
+    a = np.hypot((x + xc), (z - zc))
     k2 = 4 * x * xc / a**2
+    # Avoid NaN when coil on grid point
     k2 = clip_nb(k2, GREENS_ZERO, 1.0 - GREENS_ZERO)
-    i1 = ellipk_nb(k2) / a
-    i2 = ellipe_nb(k2) / (a**3 * (1 - k2))
-    return MU_0_2PI * ((z - zc) * (i1 - i2 * ((z - zc) ** 2 + x**2 + xc**2)))
+    return a, k2
+
+
+@nb.jit(nopython=True)
+def calc_e_k(
+    k2: float | np.ndarray,
+):
+    return ellipe_nb(k2), ellipk_nb(k2)
```

Comment:
I believe the speed gains are potentially to be had in the combined use of AGMs to calculate E and K simultaneously...

Reply:
Hi, Sorry I missed this, that is true, I separated the issues into three - AGM speed gains are in this one: https://github.com/Fusion-Power-Plant-Framework/bluemira/issues/3273).